File: src/main/mpython/py/metatron/soc/device/device.py
#  metatron: a distributed virtual machine and language
#   Copyright (C) 2025- PhaseShift Studio, LLC
# 
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU Affero General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
# 
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Affero General Public License for more details.
# 
#  You should have received a copy of the GNU Affero General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
from metatron.obj import Rec
from metatron.util.mach import router


class Device(Rec):
    def __init__(self, soc_vid, pvm: dict, tid, name: str):
        Rec.__init__(self, pvm, tid)
        self.soc_vid = soc_vid
        self.name = name

    def start(self) -> 'Device':
        return self

    def stop(self) -> 'Device':
        router().unsubscribe(self.soc_vid.extend(self.name).extend("+"))
        router().write(self.soc_vid.extend(self.name), None)

    def broadcast(self, key):
        if self.vid is not None:
            router().write(self.vid, self.pvm if key == "" else self.pvm[key])

File: src/main/java/studio/phaseshift/metatron/isa/iot/type/Device.java
/*
 * Metatron: A Distributed Computing Language and Virtual Machine
 *  Copyright (C) 2025- PhaseShift Studio, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package studio.phaseshift.metatron.isa.iot.type;

import studio.phaseshift.metatron.isa.m.type.Obj;
import studio.phaseshift.metatron.isa.m.type.Type;

import static studio.phaseshift.metatron.isa.iot.haos.haosInstSet.HAOS_ENTITY_TYPE;
import static studio.phaseshift.metatron.isa.iot.iotInstSet.DEVICE_TID;
import static studio.phaseshift.metatron.isa.m.mInstSet.REC_TID;
import static studio.phaseshift.metatron.isa.m.parser.mFluent.StartLess.isa_;
import static studio.phaseshift.metatron.isa.m.type.impl.MLst.lst;
import static studio.phaseshift.metatron.isa.m.type.impl.MRec.rec;
import static studio.phaseshift.metatron.isa.m.type.impl.MUri.uri;

/*
 * @author Marko A. Rodriguez (http://markorodriguez.com)
 */
public interface Device extends Obj {

    Type DEVICE_TYPE = Type.Builder.build()
            .tid(REC_TID)
            .vid(DEVICE_TID)
            .predicate(isa_(rec(uri("entity"), lst(HAOS_ENTITY_TYPE)))).create();


}
